src/dtx/dtx_resync.c

```diff
@@ -0,0 +1,387 @@
+			rc = dtx_resync_commit(dra->po_uuid, cont,
+					       drh, count, dra->version, block);
+			if (rc < 0)
+				err = rc;
+			count = 0;
+		}
+	}
+
+	if (count > 0) {
+		rc = dtx_resync_commit(dra->po_uuid, cont, drh, count,
+				       dra->version, block);
+		if (rc < 0)
+			err = rc;
+	}
+
+	if (layout != NULL)
+		pl_obj_layout_free(layout);
+
+	return err;
+}
+
+static int
+dtx_iter_cb(uuid_t co_uuid, vos_iter_entry_t *ent, void *args)
+{
+	struct dtx_resync_args		*dra = args;
+	struct dtx_resync_entry		*dre;
+
+	/* Ignore new DTX after the rebuild/recovery start.
+	 *
+	 * XXX: The time(NULL) based checking may be untrusted because of
+	 *	potential clock drift. We may consider to replace it with
+	 *	hybrid-clock based mechanism in future when it is ready.
+	 */
+	if (ent->ie_dtx_sec > dra->cont->sc_dtx_resync_time)
+		return 0;
+
+	/* We commit the DTXs periodically, there will be not too many DTXs
+	 * to be checked when resync. So we can load all those uncommitted
+	 * DTXs in RAM firstly, then check the state one by one. That avoid
+	 * the race trouble between iteration of active-DTX tree and commit
+	 * (or abort) the DTXs (that will change the active-DTX tree).
+	 */
+
+	D_ALLOC_PTR(dre);
+	if (dre == NULL)
+		return -DER_NOMEM;
+
+	dre->dre_xid = ent->ie_xid;
+	dre->dre_oid = ent->ie_oid;
+	dre->dre_intent = ent->ie_dtx_intent;
+	dre->dre_hash = ent->ie_dtx_hash;
+	d_list_add_tail(&dre->dre_link, &dra->tables.drh_list);
+	dra->tables.drh_count++;
+
+	return 0;
+}
+
+int
+dtx_resync(daos_handle_t po_hdl, uuid_t po_uuid, uuid_t co_uuid, uint32_t ver,
+	   bool block)
+{
+	struct ds_cont			*cont = NULL;
+	struct dtx_resync_args		 dra = { 0 };
+	struct dtx_resync_entry		*dre;
+	struct dtx_resync_entry		*next;
+	int				 rc = 0;
+	int				 rc1 = 0;
+	bool				 shared = false;
+
+	rc = ds_cont_lookup(po_uuid, co_uuid, &cont);
+	if (rc != 0) {
+		D_ERROR("Failed to open container for resync DTX "
+			DF_UUID"/"DF_UUID": rc = %d\n",
+			DP_UUID(po_uuid), DP_UUID(co_uuid), rc);
+		return rc;
+	}
+
+	while (cont->sc_dtx_resyncing) {
+		if (!block)
+			goto out;
+
+		shared = true;
+		/* Someone is resyncing the DTXs, let's wait and retry. */
+		ABT_thread_yield();
+	}
+
+	/* Some others just did the DTX resync, needs not to repeat. */
+	if (shared)
+		goto out;
+
+	cont->sc_dtx_resyncing = 1;
+	cont->sc_dtx_resync_time = time(NULL);
+
+	dra.cont = cont;
+	uuid_copy(dra.po_uuid, po_uuid);
+	dra.version = ver;
+	D_INIT_LIST_HEAD(&dra.tables.drh_list);
+	dra.tables.drh_count = 0;
+
+	D_DEBUG(DB_TRACE, "resync DTX scan "DF_UUID"/"DF_UUID" start.\n",
+		DP_UUID(po_uuid), DP_UUID(co_uuid));
+
+	rc = ds_cont_iter(po_hdl, co_uuid, dtx_iter_cb, &dra, VOS_ITER_DTX);
+
+	/* Handle the DTXs that have been scanned even if some failure happend
+	 * in above ds_cont_iter() step.
+	 */
+	rc1 = dtx_status_handle(&dra, block);
+
+	d_list_for_each_entry_safe(dre, next, &dra.tables.drh_list, dre_link)
+		dtx_dre_release(&dra.tables, dre);
```

Comment:
here seems the drh_list should be empty as all entries should be released by dtx_status_handle()? why need to traverse and free it again?

Reply:
I will change it as ASSERT() in subsequent patch.

As for the "skip_cos", our DTX CoS is "oid + dkey" based, we never cache the zeroed dkey_hash case, that is mainly for object punch case. Because the object punch potentially conflict with all the other modifications on such object, then we prefer to commit related DTX as fast as possible. That is why we do not cache zeroed dkey_hash case. Zeroed dkey_hash will be skipped lower layer CoS directly.